db/cache/charts.go

```diff
@@ -959,57 +981,76 @@ func accumulate(data ChartUints) ChartUints {
+// minus the first element is returned for convenience. limit defines the number
+// of blocks to consider for the current limitType from the best block.
+func blockTimes(blocks ChartUints, limit int) (ChartUints, ChartUints, ChartFloats) {
 	dataLen := len(blocks)
+	keys := make(ChartUints, 0, len(blocks))
 	if dataLen < 2 {
 		// Fewer than two data points is invalid for btw. Return empty data sets so
 		// that the JSON encoding will have the correct type.
-		return times, times
+		return keys, keys, newChartFloats(len(blocks))
+	}
+
+	var k int
+	if limit > 0 && limit < len(blocks) {
+		k = len(blocks) - limit
 	}
-	last := blocks[0]
-	for _, v := range blocks[1:] {
+
+	last := blocks[k]
+	k++
+	tracker := make(map[uint64]uint64)
+	for _, v := range blocks[k:] {
 		dif := v - last
 		if int64(dif) < 0 {
 			dif = 0
 		}
-		times = append(times, dif)
 		last = v
+		if _, ok := tracker[dif]; !ok {
+			keys = append(keys, dif)
+		}
+		tracker[dif]++
 	}
-	return blocks[1:], times
-}
-
-// Take the average block times on the intervals defined by the ticks argument.
-func avgBlockTimes(ticks, blocks ChartUints) (ChartUints, ChartUints) {
-	if len(ticks) < 2 {
-		// Return empty arrays so that JSON-encoding will have the correct type.
-		return ChartUints{}, ChartUints{}
-	}
-	avgDiffs := make(ChartUints, 0, len(ticks)-1)
-	times := make(ChartUints, 0, len(ticks)-1)
-	nextIdx := 1
-	workingOn := ticks[0]
-	next := ticks[nextIdx]
-	lastIdx := 0
-	for i, t := range blocks {
-		if t > next {
-			_, pts := blockTimes(blocks[lastIdx:i])
-			avgDiffs = append(avgDiffs, pts.Avg(0, len(pts)))
-			times = append(times, workingOn)
-			nextIdx++
-			if nextIdx > len(ticks)-1 {
-				break
-			}
-			lastIdx = i
-			next = ticks[nextIdx]
-			workingOn = next
+
+	var sumF, sumFx uint64
+	for x, f := range tracker {
+		sumFx += f * x
+		sumF += f
+	}
+
+	var distrMean = float64(sumFx) / float64(sumF)
+
+	var xValue = newChartUints(len(tracker))
+	var blockCount = newChartUints(len(tracker))
+	var expectedDistr = newChartFloats(len(tracker))
+	sort.Slice(keys, func(i, j int) bool { return keys[i] < keys[j] })
+
+	var lastDistr, lastTimestamp float64
+	for _, timeVal := range keys {
+		timeInterval := float64(timeVal)
+		distr := math.Exp(timeInterval * -1 / distrMean)
+		// Check if timestamp value is not consecutive from the current i.e.
+		// current = previous + 1 recompute lastDistr to ensure a smooth curve.
+		if timeInterval-lastTimestamp > 1 {
+			lastDistr = math.Exp((timeInterval - 1) * -1 / distrMean)
 		}
+
+		var expectedCount float64
+		if lastDistr-distr > 0 {
+			expectedCount = (lastDistr - distr) * float64(sumF)
+		}
+
+		xValue = append(xValue, timeVal)
+		blockCount = append(blockCount, tracker[timeVal])
+		expectedDistr = append(expectedDistr, math.Floor(expectedCount*1e2)/1e2)
+
+		lastDistr = distr
+		lastTimestamp = timeInterval
 	}
-	return times, avgDiffs
+	return xValue, blockCount, expectedDistr
```

Comment:
Let's add some unit tests for `blockTimes`, covering all the edge cases we can think of.

e.g.

```go
func Test_blockTimes(t *testing.T) {
	type args struct {
		blocks ChartUints
		limit  int
	}
	blockSeconds := ChartUints{1, 20, 33, 43, 56, 60, 79}
	tests := []struct {
		name  string
		args  args
		want0 ChartUints
		want1 ChartUints
		want2 ChartFloats
	}{
		{
			name: "empty",
			args: args{
				blocks: nil,
				limit:  0,
			},
			want0: ChartUints{},
			want1: ChartUints{},
			want2: ChartFloats{},
		},
		{
			name: "basicOK6",
			args: args{
				blocks: blockSeconds,
				limit:  6,
			},
			want0: ChartUints{4, 10, 13, 19},
			want1: ChartUints{1, 1, 2, 1},
			want2: ChartFloats{0.31, 0.18, 0.14, 0.08},
		},
		{
			name: "basicOK7",
			args: args{
				blocks: blockSeconds,
				limit:  7,
			},
			want0: ChartUints{4, 10, 13, 19},
			want1: ChartUints{1, 1, 2, 2},
			want2: ChartFloats{0.35, 0.22, 0.17, 0.11},
		},
		{
			name: "allBig",
			args: args{
				blocks: blockSeconds,
				limit:  999999,
			},
			want0: ChartUints{4, 10, 13, 19},
			want1: ChartUints{1, 1, 2, 2},
			want2: ChartFloats{0.35, 0.22, 0.17, 0.11},
		},
		{
			name: "all0",
			args: args{
				blocks: blockSeconds,
				limit:  0,
			},
			want0: ChartUints{4, 10, 13, 19},
			want1: ChartUints{1, 1, 2, 2},
			want2: ChartFloats{0.35, 0.22, 0.17, 0.11},
		},
		{
			name: "1",
			args: args{
				blocks: blockSeconds,
				limit:  1,
			},
			want0: ChartUints{},
			want1: ChartUints{},
			want2: ChartFloats{},
		},
		{
			name: "2",
			args: args{
				blocks: blockSeconds,
				limit:  2,
			},
			want0: ChartUints{19},
			want1: ChartUints{1},
			want2: ChartFloats{0.01}, // Is this right???
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, got1, got2 := blockTimes(tt.args.blocks, tt.args.limit)
			if !reflect.DeepEqual(got, tt.want0) {
				t.Errorf("blockTimes() got = %v, want %v", got, tt.want0)
			}
			if !reflect.DeepEqual(got1, tt.want1) {
				t.Errorf("blockTimes() got1 = %v, want %v", got1, tt.want1)
			}
			if !reflect.DeepEqual(got2, tt.want2) {
				t.Errorf("blockTimes() got2 = %v, want %v", got2, tt.want2)
			}
		})
	}
}
```